### DISCUSSION TITLE
Remove 'view' and 'preview' buttons for certain content types

### DISCUSSION THREAD

**Discussion by netwavebe** [OrchardCMS/OrchardCore#7002] [Q&A]
Hi,

Content types are sometimes used as an entity for holding data. They do not have a visual representation.

Yet when creating/editing them, there is a 'preview' button and in the list of content items there is a 'view' button.

Is there a way to remove these buttons for certain content types, as they might confuse end-users?

With kind regards,
Steven Spits

**Comment by ns8482e** [comment]
Shape can be removed using using placement 

```
"ContentPreview_Button": [
    {
      "place": "-",
      "contentType": [ "Menu" ]
    }
  ]
```

**Reply by aaronamm** [reply]
@ns8482e  Hi how do you hide blog content item with placement.json. I can't make it work.

Here is what I have done so far:

- Create an Orchard project with a blog recipe.
- Create a custom theme.
- Put placement.info in the root of my theme.
Here is the content of placement.info:
```json
{
"Content__Blog": [
  {
    "place": "-"
  }
],
"Content_Blog": [
  {
    "place": "-"
  }
],
"Blog": [
  {
    "place": "-"
  }
]
}
```

- None of these filters work, a blog content item still show on home page. I think to simplify example code we can remove other option, e.g. display type.
- I can create Views/Content-Blog.cshtml with empty content in my custom theme but I think it is tricky and not the right way to do it.
Please could you help me figure it out? I have tried to debug source code but still don't' understand how it work. 
From what I understand now, the shape name should be "Content__Blog".

More information:
Here is what I found from the v1 document:
"Placement works only on parts (and some fields) of content items" which turn out we can't remove content item with placement.json.
We can only remove parts or fields of a content item. An example code to remove HtmlBodyPart for Blog content type:
```json
{
  "HtmlBodyPart": [
    {
      "place": "-",
      "contentType": "Blog"
    }
  ]
}
```

**Reply by ns8482e** [reply]
@aaronamm  it may be related to #6788 

My understanding is that `place` property is used for where to place child shapes on parent shape. - Here Blog is Content Item and parent for all parts. So yes  I guess placement is working only on parts and not on content/widgets. 

As `Blog`  is being content item -  you can manage to show/hide from admin. The simple options is to change the home page to another content item or unpublish the the `Blog` content item to hide from your website.

**Comment by netwavebe** [comment]
The code above removes the 'Preview' button when editing an item, thanks!

Still looking for a way to remove the 'View' button in the list of items. This is the view that renders that button:
https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Contents/Views/Contents-Button-Edit.SummaryAdmin.cshtml#L19

Would it be possible to override Contents-Button-Edit.SummaryAdmin.cshtml for a specific content type? What would be the name of the file if the content type is 'Brand" for example?

**Reply by ns8482e** [reply]
You can override template `Contents-Button-Edit.SummaryAdmin.cshtml` in Admin theme or in Application, or in AdminTemplate by overriding `Contents_SummaryAdmin__Button__Edit` for all or by creating custom shape `Contents_SummaryAdmin__Custom__Button__Edit` to target selected content type using placement json

```
"Contents_SummaryAdmin__Button__Edit": [
    {
      "contentType": [ "Custom" ],
      "shape" : "Contents_SummaryAdmin__Custom__Button__Edit"
    }
  ]

```

**Reply by netwavebe** [reply]
@ns8482e Thanks for you feedback. This does something strange...

I used this in placement.json:

```
"Contents_SummaryAdmin__Button__Edit": [
	{
		"contentType": [ "Brand" ],
		"shape": "Contents_SummaryAdmin__Button__Edit__NoView"
	}
]
```

File `Contents-Button-Edit-NoView.SummaryAdmin.cshtml` is a copy of `Contents-Button-Edit.SummaryAdmin.cshtml` with the 'View' button removed, placed inside a custom module.

This is the output:

![image](https://user-images.githubusercontent.com/3008547/92459363-e1109380-f1c6-11ea-804b-03c6ba7b9724.png)

Any ideas?

**Reply by ns8482e** [reply]
Magic of Orchard Core Shapes! Why four times? answer is here https://github.com/OrchardCMS/OrchardCore/blob/85b4f99618ba263b55cb72e366d973e1967b2298/src/OrchardCore.Modules/OrchardCore.Contents/Drivers/ContentsDriver.cs#L51-L54 

It seems your new shape is not placed in `action` zone.  add ` "place" : "Actions:10" ` in your placement.

**Reply by netwavebe** [reply]
Adding `Actions:10` moves all 4 buttons to the right (action zone) but there are still 4 buttons...

Seems like it should only match the third shape in the list, but matches all four instead? I think I have seen something similar recently, but can't find it anymore. It was a bagpart of flowpart IIRC.

**Reply by ns8482e** [reply]
@netwavebe It seems it could be a bug. Try and see if it works with an alternate.

**Reply by netwavebe** [reply]
This is what I was thinking about: https://github.com/OrchardCMS/OrchardCore/issues/5149

Raised by you ;-)

I have no idea what alternate I should try?

**Reply by ns8482e** [reply]
@netwavebe It's bug but not same as #5149,  when you add a placement for a `shape` type  that is inside of `CombinedResult`, for some reason,  the placement info is also being applied to other shapes within `CombinedResult` 

Workaround would be to use `alternates` in placement as following.

```
"Contents_SummaryAdmin__Button__Edit": [
	{
                "place" :  "Actions:10",
		"contentType": [ "Brand" ],
		"alternates": [ "Contents_SummaryAdmin__Button__Edit__NoView" ]
	}
]
```

and in `cshtml` exclude all other `shape` types and only render for `Contents_SummaryAdmin__Button__Edit`


```cshtml
@if(Model.Metadata.Type == "Contents_SummaryAdmin__Button__Edit"){

    <!-- HTML Content Here-->

}
```

**Update**

It's due to the fact that original shape type is identified before `__`  in shape templates i.e.  original shape type for `Contents_SummaryAdmin__Tags` , `Contents_SummaryAdmin__Meta`, `Contents_SummaryAdmin__Button__Edit`, and   `Contents_SummaryAdmin__Button__Actions` is `Contents_SummaryAdmin` hence placement is applied to all. 
see, https://github.com/OrchardCMS/OrchardCore/blob/31a83f73bfef49180ebc48d77141242843924a8d/src/OrchardCore/OrchardCore.DisplayManagement/BaseDisplayManager.cs#L45
